Optimize call indirect for R2R, Arm and Arm64 scenarios

kunalspathak opened the pull request:
We generate unnecessary load of indirect cell address and can use the address that we have already loaded in ARM64. With this change, we save 8 bytes / call-site that does call indirect. The way it will be optimized is during lowering, `controlExpr` won't be created for such cases. Instead we would use temp register for such calls and during codegen, load the indirect from x11 into that temp register before calling the address in that temp register.

Address to some extent: https://github.com/dotnet/runtime/issues/35108

```
Crossgen CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit
Summary of Code Size diffs:
(Lower is better)
Total bytes of diff: -5828536 (-10.07% of base)
    diff is an improvement.
Top file improvements (bytes):
     -687316 : System.Private.CoreLib.dasm (-10.30% of base)
     -551304 : System.Private.Xml.dasm (-10.26% of base)
     -418840 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-7.36% of base)
     -415680 : Microsoft.CodeAnalysis.VisualBasic.dasm (-10.94% of base)
     -401832 : System.Linq.Expressions.dasm (-7.79% of base)
     -364704 : Microsoft.CodeAnalysis.CSharp.dasm (-10.42% of base)
     -209096 : System.Data.Common.dasm (-10.82% of base)
     -136088 : System.Private.DataContractSerialization.dasm (-10.54% of base)
     -113512 : Newtonsoft.Json.dasm (-10.62% of base)
     -109584 : Microsoft.CodeAnalysis.dasm (-8.99% of base)
     -105208 : System.Net.Http.dasm (-11.00% of base)
      -94208 : System.DirectoryServices.dasm (-13.50% of base)
      -83672 : Microsoft.VisualBasic.Core.dasm (-11.55% of base)
      -77344 : System.Management.dasm (-12.61% of base)
      -73224 : System.Linq.Parallel.dasm (-7.63% of base)
      -72856 : System.Drawing.Common.dasm (-13.98% of base)
      -72224 : System.Text.Json.dasm (-9.51% of base)
      -69048 : System.Security.Cryptography.Pkcs.dasm (-12.93% of base)
      -60840 : System.Configuration.ConfigurationManager.dasm (-10.96% of base)
      -59324 : Microsoft.CSharp.dasm (-10.35% of base)
190 total files with Code Size differences (190 improved, 0 regressed), 76 unchanged.
Top method improvements (bytes):
      -75184 (-19.99% of base) : System.Linq.Expressions.dasm - System.Linq.Expressions.Interpreter.CallInstruction:FastCreate(System.Reflection.MethodInfo,System.Reflection.ParameterInfo[]):System.Linq.Expressions.Interpreter.CallInstruction (241 methods)
       -7456 (-6.93% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ApplicationServerTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
       -6152 (-20.63% of base) : System.Private.CoreLib.dasm - HtmlEntities:.cctor()
       -5680 (-15.23% of base) : System.Text.RegularExpressions.dasm - System.Text.RegularExpressions.RegexCompiler:GenerateOneCode():this
       -4536 (-17.26% of base) : System.Data.Common.dasm - System.Data.BinaryNode:EvalBinaryOp(int,System.Data.ExpressionNode,System.Data.ExpressionNode,System.Data.DataRow,int,System.Int32[]):System.Object:this
       -4104 (-13.29% of base) : System.Private.Xml.dasm - System.Xml.Schema.XsdBuilder:.cctor()
       -3800 (-19.60% of base) : System.DirectoryServices.AccountManagement.dasm - System.DirectoryServices.AccountManagement.ADStoreCtx:.cctor()
       -3680 (-12.77% of base) : System.Management.dasm - System.Management.ManagementClassGenerator:GenerateMethods():this
       -3432 (-18.86% of base) : System.DirectoryServices.AccountManagement.dasm - System.DirectoryServices.AccountManagement.ADAMStoreCtx:.cctor()
       -3432 (-15.56% of base) : System.Management.dasm - System.Management.ManagementClassGenerator:AddToDateTimeFunction():this
       -3432 (-11.44% of base) : System.Private.CoreLib.dasm - System.Globalization.CultureData:get_RegionNames():System.Collections.Generic.Dictionary`2[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
       -3424 (-13.74% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlReflectionImporter:ImportAccessorMapping(System.Xml.Serialization.MemberMapping,System.Xml.Serialization.FieldModel,System.Xml.Serialization.XmlAttributes,System.String,System.Type,bool,bool,System.Xml.Serialization.RecursionLimiter):this
       -3224 (-13.76% of base) : System.Management.dasm - System.Management.ManagementClassGenerator:GenerateTypeConverterClass():System.CodeDom.CodeTypeDeclaration:this
       -3168 (-16.95% of base) : System.DirectoryServices.AccountManagement.dasm - System.DirectoryServices.AccountManagement.SAMStoreCtx:.cctor()
       -3128 (-8.11% of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.DesktopAssemblyIdentityComparer:.cctor()
       -3064 (-4.70% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.KernelTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
       -2800 (-16.95% of base) : System.Management.dasm - System.Management.ManagementClassGenerator:AddToDMTFDateTimeFunction():this
       -2784 (-14.20% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.MicrosoftWindowsTCPIPTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
       -2696 (-14.20% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.WpfTraceEventParser:EnumerateTemplates(System.Func`3[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Microsoft.Diagnostics.Tracing.EventFilterResponse, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]],System.Action`1[[Microsoft.Diagnostics.Tracing.TraceEvent, Microsoft.Diagnostics.Tracing.TraceEvent, Version=2.0.49.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]):this
       -2688 (-15.59% of base) : Microsoft.CodeAnalysis.CSharp.dasm - <>c:.cctor() (168 methods)
Top method improvements (percentages):
        -192 (-27.91% of base) : Microsoft.CodeAnalysis.dasm - Roslyn.Utilities.PortableShim:Initialize()
         -40 (-27.78% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberFieldSymbol:GetFieldDeclaration(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode):Microsoft.CodeAnalysis.CSharp.Syntax.BaseFieldDeclarationSyntax
         -24 (-27.27% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.FileIO.FileSystem:NormalizePath(System.String):System.String
         -24 (-27.27% of base) : System.Diagnostics.TraceSource.dasm - System.Diagnostics.Trace:Refresh()
         -88 (-26.19% of base) : System.Reflection.Metadata.dasm - System.Reflection.Metadata.Ecma335.MetadataBuilder:ValidateOrder():this
         -24 (-26.09% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicCompilation:get_RootNamespace():Microsoft.CodeAnalysis.VisualBasic.Symbols.NamespaceSymbol:this
         -24 (-26.09% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.ProjectData:EndApp()
         -24 (-26.09% of base) : System.Configuration.ConfigurationManager.dasm - System.Configuration.DateTimeConfigurationCollection:get_Item(int):System.DateTime:this
         -24 (-26.09% of base) : System.Data.Common.dasm - System.Data.DataTable:CreateInstance():System.Data.DataTable:this
         -24 (-26.09% of base) : System.Linq.Expressions.dasm - System.Dynamic.Utils.TypeUtils:IsUnsigned(System.Type):bool
         -32 (-25.81% of base) : System.Net.Mail.dasm - ConnectAndHandshakeAsyncResult:End(System.IAsyncResult)
         -16 (-25.00% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.ExplicitInterfaceHelpers:FindExplicitlyImplementedProperty(Microsoft.CodeAnalysis.CSharp.Symbols.PropertySymbol,Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,System.String,Microsoft.CodeAnalysis.CSharp.Syntax.ExplicitInterfaceSpecifierSyntax,Microsoft.CodeAnalysis.DiagnosticBag):Microsoft.CodeAnalysis.CSharp.Symbols.PropertySymbol
         -16 (-25.00% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Symbols.ExplicitInterfaceHelpers:FindExplicitlyImplementedMethod(Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol,Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol,System.String,Microsoft.CodeAnalysis.CSharp.Syntax.ExplicitInterfaceSpecifierSyntax,Microsoft.CodeAnalysis.DiagnosticBag):Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol
         -16 (-25.00% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode:get_Parent():Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode:this
         -24 (-25.00% of base) : Microsoft.CodeAnalysis.CSharp.dasm - EndInvokeMethod:get_BoundAttributesSource():Microsoft.CodeAnalysis.CSharp.Symbols.SourceMethodSymbol:this
         -24 (-25.00% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Emit.PEModuleBuilder:GetInitArrayHelper():Microsoft.Cci.IMethodReference:this
         -16 (-25.00% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode:get_Parent():Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode:this
         -24 (-25.00% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Emit.PEModuleBuilder:GetInitArrayHelper():Microsoft.Cci.IMethodReference:this
         -16 (-25.00% of base) : Microsoft.CSharp.dasm - System.Runtime.InteropServices.Variant:get_AsVariant():System.Object:this
         -16 (-25.00% of base) : Microsoft.CSharp.dasm - Microsoft.CSharp.RuntimeBinder.ComInterop.UnsafeMethods:GetObjectForVariant(System.Runtime.InteropServices.Variant):System.Object
155334 total methods with Code Size differences (155334 improved, 0 regressed), 67835 unchanged.
Completed analysis in 253.06s
```

kunalspathak commented:
@dotnet/jit-contrib , @jeffschwMSFT , @richlander , @davidwrighton , @jkotas

BruceForstall commented:
Can you show a snippet of how the generated arm32 and arm64 code changes?

kunalspathak commented:
> Can you show a snippet of how the generated arm32 and arm64 code changes?

@BruceForstall , here you go - 

ARM32:

![image](https://user-images.githubusercontent.com/12488060/81998022-4b913880-9606-11ea-84e0-064997fc4359.png)

ARM64:
![image](https://user-images.githubusercontent.com/12488060/81998054-5ea40880-9606-11ea-8dc9-af257a7f2b16.png)

kunalspathak commented:
@BruceForstall , @CarolEidt - can you please take another look at the fix for one of the failing test? There was a JIT helper call that was dead and it was getting removed. However one of the argument node to the call that represented the address and whose calculation I optimized as part of this PR was not getting removed. This lead to generating code for the left over node that ended up reading the value from memory address present in `x11`. `x11` was hold `0x00` and I was getting `AccessViolation`. The fix was to remove the left over argument to the helper call. This could have been done in `fgComputeLifeLIR` but I felt the more natural place is to handle it inside `fgAddrCouldBeNull`. Here we will check if the node holding the address is `GT_PHYSREG` and if yes, check if it is `REG_R2R_INDIRECT_PARAM` so we know that the address in that register cannot be `null`. This would let us remove the optimized argument node as well.

BruceForstall commented on src/coreclr/src/jit/flowgraph.cpp:
```diff
@@ -7417,6 +7417,14 @@ bool Compiler::fgAddrCouldBeNull(GenTree* addr)
+    {
+        // Just to be conservative, ensure the srcReg is the known REG_R2R_INDIRECT_PARAM
+        // if we are claiming that addr cannot be null.
+        return addr->AsPhysReg()->gtSrcReg != REG_R2R_INDIRECT_PARAM;
```
This doesn't seem safe to me. It's saying that r11 as a GT_PHYSREG can never be null. That might be true in the case where r11 is loaded as a r2r indirect special argument, but r11 can be used for other purposes, and this function might (eventually if not now) be used for other purposes as well, in which case this wouldn't always hold.

kunalspathak replied:
Got it. In that case, do you think a better place would be inside `fgComputeLifeLIR()`?

BruceForstall replied:
Yes, it feels like someplace inside `Removing dead call` code might be appropriate, but I'm not confident about that.

kunalspathak replied:
Yes, I tried marking the node to be unused at place that you suggest (inside code where we iterate over the arguments of the call to see if they can be marked unused). but I hit assert going forward inside [lir](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lir.cpp#L1623). Instead what I was thinking was to add a case for `GT_IND` [here](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/liveness.cpp#L2035) and the code will be exactly same as `default case` but with added condition that checks if the node is `GT_PHYSREG`  and if yes, remove it.

kunalspathak replied:
Ok, made the fix in liveness.

BruceForstall commented on src/coreclr/src/jit/emitarm.cpp:
```diff
@@ -7676,7 +7676,7 @@ void emitter::emitInsLoadStoreOp(instruction ins, emitAttr attr, regNumber dataR
     GenTree* addr = indir->Addr();
 
-    if (addr->isContained())
+    if (addr->isContained() && !addr->OperIs(GT_PHYSREG))
```
Can you explain why you couldn't use the form `IND(LEA(PHYSREG r11))` and then not need these special cases, instead of using `IND(PHYSREG r11)` that requires the special case? In both cases, the physreg would get marked contained, and the LEA would also be marked contained.

kunalspathak replied:
From what I understand, `PHYSREG` still assigns a new register (in this case `x4`) and not reuse `x11` because it is occupied. Hence we generate extra `mov x4, x11`.

```
                                                              /--*  t16    long   
Generating: N025 (???,???) [000030] ------------        t30 = *  PUTARG_REG long   REG x11
Generating: N027 (  1,  1) [000031] ------------        t31 =    PHYSREG   long   x11 REG x4
IN0007:                           mov     x4, x11
                                                              /--*  t31    long   
Generating: N029 (  2,  2) [000032] -c----------        t32 = *  LEA(b+0)  long   REG NA
                                                              /--*  t32    long   
Generating: N031 (  5,  4) [000033] ------------        t33 = *  IND       long   REG x4
IN0008:                           ldr     x4, [x4]
```

BruceForstall commented on src/coreclr/src/jit/liveness.cpp:
```diff
@@ -2040,7 +2040,19 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
+                        GenTree* addr = node->AsIndir()->Addr();
+                        if (addr->OperGet() == GT_PHYSREG)
+                        {
+                            isR2RIndirectParam = addr->AsPhysReg()->gtSrcReg == REG_R2R_INDIRECT_PARAM;
```
It looks to me like this code will eliminate any `IND(PHYSREG x11)`. It's possibly the case that the only way these are getting introduced today is in your change to lower.cpp. But is it safe, long-term? Not every use of x11 is a R2RIndirectParam.

It seems like you need to delete only, and exactly, the cases where these are added as a special call argument. Wouldn't that be in the `GT_CALL` case above, where it's looking through the arguments of a call that it has already decided to remove?

kunalspathak committed:
Use a different approach to optimize the indirect calls for R2R

During lowering, don't create a controlExpr for indirect call. Instead use
temp register for such calls and during codegen, load the indirect from x11
into that temp register before calling the address in that temp register.

CarolEidt commented on src/coreclr/src/jit/lower.cpp:
```diff
@@ -3284,9 +3284,25 @@ GenTree* Lowering::LowerDirectCall(GenTreeCall* call)
+#if defined(FEATURE_READYTORUN_COMPILER) && defined(TARGET_ARMARCH)
+            if (call->IsR2RRelativeIndir())
+            {
+                cellAddr = PhysReg(REG_R2R_INDIRECT_PARAM);
```
I think it's a mistake to use a `GT_PHYSREG` here. IMO we should eliminate those, as their semantics are unclear.
I think the right answer is to return `nullptr` in this case. Then, in `BuildCall`, you would request an internal register for this case (`buildInternalIntRegisterDefForNode(call)`) as is currently done on Arm32 for the null `ctrlExpr` case. This register would then be used in codegen to load from `X11` and use as the call target.

kunalspathak replied:
Yes, I chatted with @BruceForstall  yesterday and trying out https://github.com/dotnet/runtime/pull/36712. If you don't mind please have a look into that PR to make sure I captured it correctly.

kunalspathak committed:
review feedback

kunalspathak commented:
@jaredpar - Is these CI failures known issue? I am seeing bunch of these:
```
##[error]Path does not exist: /Users/runner/runners/2.168.2/work/1/s/artifacts/log
```

jaredpar commented:
@kunalspathak all of the checks passed here. Where were you seeing that error?

CarolEidt reviewed (approved):
LGTM - thanks!

kunalspathak commented:
> @kunalspathak all of the checks passed here. Where were you seeing that error?

Yeah, they were failing earlier but passed after rerun.

BruceForstall reviewed (approved):
LGTM. Glad this option worked out.

kunalspathak commented:
> LGTM. Glad this option worked out.

Thanks for helping! :)
